### DISCUSSION TITLE
Msprime multiple roots?

### DISCUSSION THREAD

**Discussion by NewAtl** [tskit-dev/msprime#2301] [Q&A] [answered]
Hi everyone! 

I am working on a project that involves simulating methylation states based on the recombination patterns generated by `msprime`. My objective is to analyze the output to understand how recombination affects methylation across the genome. I'm still very new to the program, so I would appreciate some guidance :D

### Problem

I am trying to read the `nodes` and `edges` tables from `msprime` to analyze the tree structures they represent. However, I am encountering an issue:

- **Issue**: Despite setting various parameters, I consistently end up with multiple trees in the output tables. This means that:
  - I identify the root of the tree at the earliest time point present in the `nodes` table.
  - However, a new root appears at a later time point, which is not connected to the initial tree formed under the earliest root.

### Question

I believe I am misunderstanding some aspect of how `msprime` handles tree sequences or how I should interpret them. Specifically, I am confused about:

- Why do multiple roots (trees) appear in the output?
- How can I correctly interpret or set parameters to avoid this situation?

Thank you for your help! If you need more information about the parameters I am using or specific examples from the output, please let me know.

This is one of my tries to get the needed tables:
`ts = msprime.sim_ancestry(
    10, recombination_rate=0.5, sequence_length=20,additional_nodes=msprime.NodeType.RECOMBINANT,coalescing_segments_only=False
)`

**Comment by molpopgen** [comment]
Due to recombination, the root can change along the genome. In other words, different segments of the genome can have different MRCA nodes.  But with msprime, each tree will have one root:

```
>>> ts = msprime.sim_ancestry( 10, recombination_rate=0.5, sequence_length=20,additional_nodes=msprime.NodeType.RECOMBINANT,coalescing_segments_only=False )
>>> for t in ts.trees():
...     print(t.num_roots)
... 
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
1
```

**Reply by NewAtl** [reply]
Thanks for the response! Is there a way to get only one single tree instead of many?

**Reply by molpopgen** [reply]
set the recombination rate to zero.

**Reply by NewAtl** [reply]
Okay but then there isnt happening any recombination right? does recombination automatically mean that i have multiple trees with multiple roots?

**Reply by molpopgen** [reply]
Yes, a recombination rate of 0 means that there will be no recombination.  A recombination rate > 0 will lead to a random number of trees >= 0.  If the rate is low, there may be no topology changes in the history of the sample (1 tree).  As the rate increases, there are likely to be many topology changes (> 1 tree).

**Reply by molpopgen** [reply]
To clarify something: with nonzero recombination you are likely to have a tree sequence that contains trees with different roots.  Any given tree in the sequence will have 1 root. (In a "standard" coalescent simulation scenario like what msprime does.)

It is also possible for a tree sequence to have many trees, each of which has the same root.  This would mean that there are topology changes among subtrees but the MRCA node is constant along the "genome".

**Reply by NewAtl** [reply]
okay i get it. Are the tables nodes and edges always representing the tree sequence?

**Reply by molpopgen** [reply]
Yes.  You may want to take a look at the tutorial documents at tskit.dev for an overview of the data structures, etc..

**Reply by NewAtl** [reply]
Thank you! What i basically need is the ancestral recombination graph which has one root(?) where i can iterate trough it. Then i would check for the recombination events using the flags. Is there any way in msprime to get such a datastructure?

**Reply by molpopgen** [reply]
Okay, that is a rather different question! At this point I'll have to refer you to the docs/tutorials, which cover what support exists for simulating "full ARGs".

**Reply by molpopgen** [reply]
For example, see [here](https://tskit.dev/tutorials/args.html) for some inspiration.

**Reply by NewAtl** [reply]
Ah thank you! Do i understand that correct, that witht the argument `record_full_arg=True` the tables from msprime are representing an ancestral recombination graph? wouldnt that be the graph with only one root?

**Reply by molpopgen** [reply]
yes -- at this point, it may be most useful to recreate the examples from that tutorial in, for example, a jupyter notebook, to help visualize the relationship b/w the options and the output.

**Reply by jeromekelleher** [reply, marked as answer]
> wouldnt that be the graph with only one root?

The graph doesn't have to have one root though (in fact won't most of the time), because you will have different MRCA nodes at different points along the sequence. To have one root you would need to simulate back to the "Grand MRCA", which takes exponential time. 

See the "Big ARG and little ARG" section in our recent [Genetics paper](https://academic.oup.com/genetics/advance-article/doi/10.1093/genetics/iyae100/7714980) this point, and for general background on the difference between the ARG as a graph, and a succinct tree sequence. It's quite subtle, so I wouldn't expect to understand immediately.

**Reply by molpopgen** [reply]
Thanks @jeromekelleher. @NewAtl -- I didn't look closely enough at the examples in the ARG tutorial and I missed that there is a mouse-over action that shows you each embedded marginal tree.
